### PR TITLE
Allow attribute redefinition in a subclass

### LINKED ISSUE #27: Allow attributes override, or at least constraint
In DDD it may be needed to override attribute in a sub-struct. 

For example, I have a `Customer` with optional `#address`, but in `Order` I need `Customer` with mandatory `#address`. 

I have an option to define `class CustomerWithAddress < Customer` with overwritten `#address` attribute, but with current implementation it's impossible to do it. 

Also, we can think about better specific DSL for such things, like `Customer.constrained(required_attribute: :address)`. 
In this case, we need to have ability to revert `.optional` from Dry::Type.

Tell me your idea/opinion, and I can try to make PR.

### PR DESCRIPTION
**Pull request by flash-gordon**
fixes #27

### CODE REVIEW

File: lib/dry/struct/class_interface.rb [hunk 1]
```diff
@@ -67,7 +67,7 @@ def constructor_type(type = nil)
 
       def schema
         super_schema = superclass.respond_to?(:schema) ? superclass.schema : {}
-        super_schema.merge(@schema || {})
+        super_schema.merge((defined? @schema) ? @schema : {})
```
Comment (maximderbin) [inline lib/dry/struct/class_interface.rb]: I guess `super_schema.merge(defined?(@schema) ? @schema : {})` will be cleaner :)

Comment (backus) [inline lib/dry/struct/class_interface.rb]: I am pretty sure @schema is guaranteed to be set @flash-gordon. See https://github.com/dry-rb/dry-struct/blob/6ade17173ea2e7f631007aa7f8b5e85613f0de9f/lib/dry/struct/class_interface.rb#L17

**Comment by backus** [comment]
@flash-gordon if we went with the `redefine_attribute` recommendation I made in #28 then `check_schema_duplication` could remain pure I think

**Comment by timriley** [comment]
This looks good to me. As for @backus' suggestion, `redefine_attribute` feels like a departure from the typical "override behaviour a subclass" behaviour of Ruby, which feels like another violation of POLS, which is what we wanted to avoid here in the first place.

(On the other hand, it does require the class authors to be clear about their types, which is perhaps a good thing given the nature of this gem.)

Weighing both options together though, I think I lean towards what we have here, just allowing `attribute` to override an attribute in a superclass.

**Comment by backus** [comment]
> redefine_attribute feels like a departure from the typical "override behaviour a subclass" behaviour of Ruby, which feels like another violation of `POLS`, which is what we wanted to avoid here in the first place.

Adding explicit type checks for every member of an object also departs pretty significantly from normal Ruby. I think if the error raised for overwriting with `attribute` said `"If you want to overwrite an attribute, use `redefine` instead"` then it wouldn't be bad at all. Having a strict API for changing the type of an attribute seems totally reasonable for a gem where the primary use is being careful about types.

Another option could be to have `attribute ..., ..., overwrite: true`.

**Comment by flash-gordon** [comment]
@backus oh, I agree, we're really pushing Ruby to the limits, but I think in another direction :) We don't try to fix Ruby's OO approaches. I did what I got used to see in the language. In, let's say, Java, I'd did something different, I swear 😆 Also I can't argue with your point about having a meaningful error description will resolve possible complications. Let's see what @solnic says, then choose 😬

File: lib/dry/struct/class_interface.rb (hunk 1 above)
Comment (flash-gordon) [inline lib/dry/struct/class_interface.rb]: @backus nice catch, so we don't need `||` either!

**Comment by Kukunin** [comment]
We still keep constraint to avoid redeclaration within same class, but allow overriding in sub-classes. And it's pretty expected behaviour, without all `redefine` and `overwrite: true`. 

I vote for current implementation (if my vote worth something, of course =) LGTM.

**Comment by Kukunin** [comment]
btw, I believe https://github.com/dry-rb/dry-struct/blob/66884ad16bacb27bcc41b714f3ab6e0d99297a74/lib/dry/struct/class_interface.rb#L42 should be `prev_schema = @schema`, because it copies whole schema from superclass and assigns to subclass after first call of `attributes`.

**Comment by flash-gordon** [comment]
@Kukunin and why it shouldn't do so?

**Comment by Kukunin** [comment]
because `#schema` method already merges current schema with superclass's schema. 

```ruby
class A < Dry::Struct
  attribute :super, Types::String
end
class B < Dry::Struct
  attribute :sub, Types::String
end
A.instance_variable_get('@schema').keys
# => [:super]
B.instance_variable_get('@schema').keys
# => [:super, :sub]
B.schema # merges [:super] schema with [:super, :sub]
```

I admit, that it's not significant, but making`@schema` contain only attributes from subclass seems logical. I believe it won't break anything as well. 
Otherwise, we could just copy schema on `inherited` hook without any complexity.

**Comment by timriley** [comment]
I feel like the approach implemented here is still good. Using the same method for both defining/redefining is consistent with standard Ruby behaviour and also keeps our API simpler/more flexible.

**Comment by solnic** [comment]
merged 😅

**Comment by Kukunin** [comment]
that's huge change for DDD =)
will you release new version soon?)

**Comment by solnic** [comment]
@Kukunin yes I'll release it later this week

**Comment by solnic** [comment]
@Kukunin just pushed dry-struct 0.2.0

**Pull request merged**